build: run cctests as part of test-ci target

bnoordhuis opened the pull request:
For discussion.

CI: https://ci.nodejs.org/job/node-test-pull-request/3591/

bnoordhuis commented:
Different take that moves merging of the TAP files out of tools/test.py and into a standalone script: https://ci.nodejs.org/job/node-test-pull-request/3592/

richardlau commented on tools/merge-tap.py:
```diff
@@ -0,0 +1,18 @@
+      head, rest = rest.split('\n', 1)
+    if head.startswith('1..'):
+      total += int(head[3:])
+      output += rest
```
Will this work without renumbering the individual tests in the files being merged? 
As it stands I think we'll end up with multiple `ok 1 ...`.

jbergstroem commented on Makefile:
```diff
@@ -201,9 +201,11 @@ test-ci-js:
+	$(PYTHON) tools/test.py $(PARALLEL_ARGS) -p tap --logfile jstest.tap \
 		--mode=release --flaky-tests=$(FLAKY_TESTS) \
 		$(TEST_CI_ARGS) $(CI_NATIVE_SUITES) $(CI_JS_SUITES)
+	$(PYTHON) tools/merge-tap.py test.tap cctest.tap jstest.tap
```
Do we really have to merge these? We can just tell Jenkins to pick all three up.

bnoordhuis replied:
I thought that might be the case.  I'll remove the script again, it was really just to sanity-check the gtest changes.

jbergstroem replied:
Cool. Hopefully we can just glob from Jenkins. I'll give It a go.

jbergstroem commented:
cc @nodejs/build

joaocgreis commented:
On Windows, `vcbuild test-ci` already runs `cctest` but ignores the results. Here is the error from a Debug build:
```
Running main() from gtest_main.cc
[==========] Running 22 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 4 tests from UtilTest
[ RUN      ] UtilTest.ListHead
[       OK ] UtilTest.ListHead (0 ms)
[ RUN      ] UtilTest.StringEqualNoCase
[       OK ] UtilTest.StringEqualNoCase (0 ms)
[ RUN      ] UtilTest.StringEqualNoCaseN
[       OK ] UtilTest.StringEqualNoCaseN (0 ms)
[ RUN      ] UtilTest.ToLower
[       OK ] UtilTest.ToLower (0 ms)
[----------] 4 tests from UtilTest (0 ms total)

[----------] 18 tests from InspectorSocketTest
[ RUN      ] InspectorSocketTest.ReadsAndWritesInspectorMessage
Assertion failed: backlog > 0, file src\win\tcp.c, line 561
```

bnoordhuis commented:
Removed the script again, made changes to vcbuild.bat and added a -Wformat fix for the inspector cctest because I didn't feel like filing a separate PR for that. :-)

New CI: https://ci.nodejs.org/job/node-test-pull-request/3609/

@ofrobots @eugeneo Running the cctests seems to have unearthed some issues.  Several of the buildbots fail like this (from https://ci.nodejs.org/job/node-test-commit-linux/4529/nodes=debian8-x86/console): 
```
# Value of: 0
# Expected: uv_is_active(reinterpret_cast<uv_handle_t*>(&socket))
# Which is: 1
not ok 9 - InspectorSocketTest.ExtraLettersBeforeRequest
  ---
  duration_ms: 0.001
  ...
# Value of: 0
# Expected: uv_is_active(reinterpret_cast<uv_handle_t*>(&socket))
# Which is: 1
not ok 10 - InspectorSocketTest.RequestWithoutKey
  ---
  duration_ms: 0
  ...
# Value of: 0
# Expected: uv_is_active(reinterpret_cast<uv_handle_t*>(&socket))
# Which is: 1
ok 11 - InspectorSocketTest.KillsConnectionOnProtocolViolation
```

bnoordhuis commented:
Rebased, let's see if the test failures have been resolved: https://ci.nodejs.org/job/node-test-pull-request/3809/

eugeneo commented:
I'm working on it - failures are still there...

On Tue, Aug 23, 2016 at 12:10 PM Ben Noordhuis <notifications@github.com>
wrote:

> Rebased, let's see if the test failures have been resolved:
> https://ci.nodejs.org/job/node-test-pull-request/3809/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/8034#issuecomment-241843456>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARkrScGJo88I6TjfrW1w9jXvQZkVG6Fks5qi0WggaJpZM4JgM5v>
> .
>

eugeneo commented:
FYI - the test on Windows hit https://github.com/nodejs/node/issues/8155, I'll create a PR once I fix it.

bnoordhuis commented:
Rebased.  New CI run: https://ci.nodejs.org/job/node-test-pull-request/4007/

bnoordhuis commented:
@eugeneo cctest failures still appear to be unfixed:
```
not ok 8 - InspectorSocketTest.ExtraTextBeforeRequest
  ---
  duration_ms: 0
  ...
# Value of: 0
# Expected: uv_is_active(reinterpret_cast<uv_handle_t*>(&socket))
# Which is: 1
not ok 9 - InspectorSocketTest.ExtraLettersBeforeRequest
  ---
  duration_ms: 0
  ...
# Value of: 0
# Expected: uv_is_active(reinterpret_cast<uv_handle_t*>(&socket))
# Which is: 1
not ok 10 - InspectorSocketTest.RequestWithoutKey
  ---
  duration_ms: 0
  ...
# Value of: 0
# Expected: uv_is_active(reinterpret_cast<uv_handle_t*>(&socket))
# Which is: 1
```

eugeneo commented:
cctest works for me on Windows, Mac and Linux after I apply nodejs/node#8528

bnoordhuis commented:
CI with fixes from #8505 and #8528 incorporated: https://ci.nodejs.org/job/node-test-pull-request/4044/

bnoordhuis commented:
Rebased.  New CI: https://ci.nodejs.org/job/node-test-pull-request/4243/

bnoordhuis commented:
Quite a few flakes but none related to this PR.  Can I get a LGTM?

ofrobots commented:
Do you think it is possible to make/request the gtest changes upstream? There is also this: https://github.com/kinow/gtest-tap-listener.

bnoordhuis commented:
Filed https://github.com/google/googletest/issues/887 to see if they are interested in accepting this upstream.

> There is also this: https://github.com/kinow/gtest-tap-listener.

Yes, I've seen that but I don't know; I like my software with fewer moving parts, not more.

bnoordhuis commented:
Rebased, new CI: https://ci.nodejs.org/job/node-test-pull-request/4366/

@ofrobots Any objections to landing this?  I didn't get a reply from upstream.

ofrobots commented:
No objections
On Mon, Oct 3, 2016 at 7:08 AM Ben Noordhuis <notifications@github.com>
wrote:

> Rebased, new CI: https://ci.nodejs.org/job/node-test-pull-request/4366/
>
> @ofrobots <https://github.com/ofrobots> Any objections to landing this? I
> didn't get a reply from upstream.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/8034#issuecomment-251080785>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE0qVDbIW4j2k06ntKFOn2TF-HPztN5ks5qwOIigaJpZM4JgM5v>
> .
>

bnoordhuis commented:
Back-porters, commit b3d283a (the last one) should be omitted when back-porting to v4.x (but not v6.x.)

mscdex commented:
Why did the *default* result printer switch from pretty print to TAP? I get TAP output from cctest when running `make test` locally.

TheAlphaNerd commented:
waiting to backport this until we have consensus on the tap output... has this been decided on master yet?

mscdex commented:
@TheAlphaNerd The default reporter format change was reverted in master.

TheAlphaNerd commented:
@mscdex can you give me the PR that happened in?

mscdex commented:
@TheAlphaNerd https://github.com/nodejs/node/pull/8948

TheAlphaNerd commented:
@bnoordhuis or @mscdex would one of you be willing to backport this with the commit from #8948?

jbergstroem commented:
and on top, #9262.

mscdex commented:
@TheAlphaNerd https://github.com/nodejs/node/pull/9682
